api/src/main/java/com/velocitypowered/api/event/query/QueryEvent.java

```diff
@@ -0,0 +1,297 @@
+package com.velocitypowered.api.event.query;
+
+import com.velocitypowered.api.proxy.config.ProxyConfig;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.net.InetAddress;
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * This event is fired if proxy is getting queried over GS4 Query protocol
+ */
+public final class QueryEvent {
```

Comment:
Please follow the practice used in `ProxyPingEvent`: an immutable `QueryResponse` object with an associated builder.